pandas/core/internals/managers.py

```diff
@@ -1949,7 +1959,7 @@ def _check_comparison_types(
     else:
         op = np.vectorize(
             lambda x: bool(re.search(b, x))
-            if isinstance(x, str) and isinstance(b, str)
+            if isinstance(x, str) and isinstance(b, (str, Pattern))
```

Comment:
`b` is typed as `Scalar`, where
`Scalar = Union[PythonScalar, PandasScalar]`

but `b` resolves to `Union[builtins.str, builtins.float, Any]` due to unresolved imports and hence mypy is green

but should probably update types and docstring anyway

Reply:
Thanks @simonjayhawkins. Would the proper type for b be Union[Scalar, Pattern], or should PythonScalar be extended to include Pattern?

Reply:
I would keep PythonScalar unchanged for now.

Reply:
Updated. Odd that b is typed as scalar but there's an isinstance check for np.ndarray inside the function, maybe that's not needed.

Reply:
the docstring also implies b could be an array. I guess could add ArrayLike while here or leave. either works for me.

Reply:
I'll add it. Actually a can also be a scalar so the type shouldn't only be ArrayLike I don't think:
```python
In [17]: b
Out[17]: array(['a', 'a', 'a'], dtype='<U1')

In [18]: _compare_or_regex_search("a", b)
Out[18]: array([ True,  True,  True])
```

Reply:
it looks like the types were originally this way when added, see also https://github.com/pandas-dev/pandas/pull/32890#discussion_r406393255, but changed in #33598 

maybe best to keep the changes here minimal for now as this PR is being backported

(and separate PR for ensuring the types are correct and consistent with docstrings.)